ndk-build/src/manifest.rs

```diff
@@ -1,258 +1,345 @@
-            {activity_metadatas}
-            <intent-filter>
-                <action android:name="android.intent.action.MAIN" />
-                <category android:name="android.intent.category.LAUNCHER" />
-            </intent-filter>
-            {intent_filters}
-        </activity>
-    </application>
-</manifest>"#,
-            package_name = &self.package_name,
-            package_label = &self.package_label,
-            version_name = &self.version_name,
-            version_code = self.version_code,
-            split = split,
-            target_sdk_version = self.target_sdk_version,
-            min_sdk_version = self.min_sdk_version,
-            opengles_version = opengles_version,
-            target_name = &self.target_name,
-            icon = icon,
-            fullscreen = fullscreen,
-            orientation = orientation,
-            launch_mode = launch_mode,
-            application_metadatas = application_metadatas.join("\n"),
-            activity_metadatas = activity_metadatas.join("\n"),
-            debuggable = self.debuggable,
-            features = features.join("\n"),
-            permissions = permissions.join("\n"),
-            intent_filters = intent_filters.join("\n"),
-        )
+fn serialize_features<S>(features: &Option<Vec<Feature>>, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    // Check if features contains an `opengles_version` if not, add a default feature.
+    match features {
+        Some(features) => {
+            if features.iter().all(|f| f.opengles_version.is_none()) {
+                let mut f = features.clone();
+                f.push(Feature {
+                    name: None,
+                    required: Some(true),
+                    opengles_version: Some((3, 1)),
+                });
+                return serializer.serialize_some(&f);
+            }
+            serializer.serialize_some(features)
+        }
+        None => serializer.serialize_none(),
     }
+}
 
+impl AndroidManifest {
     pub fn write_to(&self, dir: &Path) -> Result<(), NdkError> {
         let mut file = File::create(dir.join("AndroidManifest.xml"))?;
-        writeln!(file, "{}", self.to_string())?;
+        let serialized = quick_xml::se::to_string(&self).unwrap();
+        writeln!(file, r#"<?xml version="1.0" encoding="utf-8"?>"#)?;
+        writeln!(file, "{}", serialized)?;
         Ok(())
     }
 }
 
-#[derive(Debug)]
-pub struct Feature {
-    pub name: String,
-    pub required: bool,
+/// See https://developer.android.com/guide/topics/manifest/application-element
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct Application {
+    #[serde(rename(serialize = "android:debuggable"))]
+    pub debuggable: Option<bool>,
+    #[serde(rename(serialize = "android:theme"))]
+    pub theme: Option<String>,
+    #[serde(rename(serialize = "android:hasCode"))]
+    #[serde(default = "default_has_code")]
+    pub has_code: Option<bool>,
```

Comment:
```suggestion
    #[serde(default)]
    pub has_code: bool,
```
That defaults to `false` as well, right?